android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/ConsentManager.java

```diff
@@ -0,0 +1,268 @@
+
+/**
+ * Manage's participant's consent to research.
+ * <p>
+ * If the participant has not given their consent for all the consents required by the study, Bridge
+ * functionality will be limited. Many APIs will return a
+ * {@link org.sagebionetworks.bridge.rest.exceptions.ConsentRequiredException} if any required
+ * consent has not been granted by the participant.
+ * <p>
+ * FIXME: handle async upload of consents which are being stored locally
+ */
+public class ConsentManager {
+    private static final Logger LOG = LoggerFactory.getLogger(ConsentManager.class);
+
+
+
+    private final AuthenticationManager authenticationManager;
+    private final ForConsentedUsersApi forConsentedUsersApi;
+    private final ConsentDAO dao;
+
+
+    public ConsentManager(@NonNull AuthenticationManager authenticationManager, ConsentDAO dao) {
+        this.authenticationManager = authenticationManager;
+        this.forConsentedUsersApi = authenticationManager.getApi();
+        this.dao = dao;
+    }
+
+    /**
+     * @param subpopulationGuid guid for the subpopulation of the consent
+     * @return true if consent to participate was made for this consent
+     */
+    public boolean isConsented(@NonNull String subpopulationGuid) {
+        checkNotNull(subpopulationGuid);
+
+        return isConsentedInSessionOrLocal(authenticationManager.getUserSessionInfo(), subpopulationGuid);
+    }
+
+    /**
+     * @param subpopulationGuid guid for the subpopulation of the consent
+     * @return true if the consent to participate was made against the most recently published
+     * version of this consent
+     */
+    public boolean isConsentedMostRecent(@NonNull String subpopulationGuid) {
+        checkNotNull(subpopulationGuid);
+
+        ConsentStatus consentStatus = getConsentStatus(subpopulationGuid);
+        if (consentStatus == null) {
+            return false;
+        }
+        return consentStatus.getSignedMostRecentConsent();
+    }
+
+    @Nullable
+    private ConsentStatus getConsentStatus(@NonNull String subpopulationGuid) {
+        checkNotNull(subpopulationGuid);
+
+        UserSessionInfo userSessionInfo = authenticationManager.getUserSessionInfo();
+        return userSessionInfo == null ? null : userSessionInfo.getConsentStatuses()
+                .get(subpopulationGuid);
+    }
+
+    // if the participant's session indicates consent to this subpopulation, use that. otherwise,
+    // treat presense of consent in DAO as having consented
+    private boolean isConsentedInSessionOrLocal(UserSessionInfo session, String subpopulationGuid) {
+        ConsentStatus subpopulationStatus = getConsentStatus(subpopulationGuid);
+        if (subpopulationStatus != null && subpopulationStatus.getConsented()) {
+            return true;
+        }
+        return dao.get(subpopulationGuid) != null;
+    }
+
+    /**
+     * @return true if all required consents have been signed
+     */
+    public boolean isConsented() {
```

Comment:
Right now mPower doesn't handle the cases associated with the way consents actually work on Bridge, so maybe I could just keep isConsented() for now, or rename to clarify.

Background:

One simplification made in MoleMapper/ResearchStack was assuming one consent, which by convention on Bridge, has a subpopulationGuid = studyId

In reality, a study can have multiple consents, with each consent associated with a subpopulationGuid, each consent can have multiple versions, and be required or not required.

So isConsented() actually needs to look through all required consents and check that the user has consented (to reflect what Bridge believes about whether the user is consented).

isConsented(subpopulation) lets you check for a specific consent

isConsentMostRecent variants takes version into account. I'd imagine this could be useful if the user signed an older version, and therefore has access to Bridge APIs, but governance would really want them to agree to the new version (kinda like new Terms and Conditions for software?)